source/common/network/socket_interface.h

```diff
@@ -52,5 +52,23 @@ static inline const SocketInterface* socketInterface(std::string name) {
 using SocketInterfaceSingleton = InjectableSingleton<SocketInterface>;
 using SocketInterfaceLoader = ScopedInjectableLoader<SocketInterface>;
 
+/**
+ * Create IoHandle for given address
+ * @param type type of socket to be requested
+ * @param addr address that is gleaned for address type, version and socket interface name
+ * @return @ref Network::IoHandlePtr that wraps the underlying socket file descriptor
+ */
+static inline IoHandlePtr ioHandleForAddr(Socket::Type type,
+                                          const Address::InstanceConstSharedPtr addr) {
+  auto sock_interface_name = addr->socketInterface();
+  if (!sock_interface_name.empty()) {
+    auto sock_interface = socketInterface(sock_interface_name);
+    RELEASE_ASSERT(sock_interface != nullptr,
```

Comment:
Regarding this RELEASE_ASSERT, what guarantees that the interface names returned by the address are valid?

Reply:
By default, the names are null. If anybody sets a name on the address, e.g., resolvers, the expectation is that those names are valid. The alternative would be to silently accept errors and use the default `SocketInterface` implementation but that would just hide misconfiguration. Still, we can change it if need be!

Reply:
We need config validation so this doesn't turn into a runtime failure at some indeterminate time after a configuration that refers to an unknown socket interface provider is specified.

Crashing the proxy in a situation like this doesn't seem like the right thing.  Terminating the connection with some critical errors may be an option, but really - we need some protections during config load to provide some better guarantees that this won't happen.

Reply:
Currently, there's no way to configure a socket interface name on an address via config. The addresses have resolvers and resolvers spawn `Address::Instance`s, with or without a socket interface name. So then, if the resolvers are misconfigured and generate broken addresses, what should we do? 

But, let's assume we'll allow configuration of addresses with socket interface names and don't validate those in resolvers or wherever we spawn `Address::Instance`s. Then, we could return nullptr `IoHandlePtr`s or an `IoSocketHandleImpl` with a -1 fd. Consequently, any place that constructs a `Socket` must afterwards check if the internal `IoHandle` was correctly initialized. A quick code check shows that `ConnectionImpl` `RELEASE_ASSERT`s that the socket it was given is valid in the constructor and similarly `NetworkListenSocket` `RELEASE_ASSERT`s that the `IoHandle`'s fd is valid.  

Do we any other alternatives?

Reply:
That's an interesting idea! It's definitely doable, assuming all the places that construct `Address::Instance`s have access to the `socket_interface.h` and therefore ability to lookup the `SocketInterface` implementations. But I'll defer this to @yanavlasov as I'm sure he has more experience with this type of problems given #11618